### DISCUSSION TITLE
Pass shape property to  URL?

### DISCUSSION THREAD

**Discussion by GanizaniSitara** [jgraph/drawio#2775] [Ideas]
As the title says, is there any way we could pass shape property to URL?

What I'd like to achieve - imagine a large enterprise (upwards of 50K Draw.IO diagrams in a single confluence instance), adopting C4. Naming convention/application catalogue not always consistent. If I can take shape property and pass it to a linked URL I could have a resolution service running that would always pass the clicker to the "knowledge repository" that knows the most wherever that lives.  

So if my C4 shape says "Application A" in a description I can resolve "Application A" to "Official name for A" and pass them to wherever they needo go. Equally "Bob" which is another name for "Application A" would also in my service resolve to "Official name for A" and pass people wherever they need to go to. 

The idea is we encode colloquial knowledge and still enable entity resolution and seamlessly link from diagrams.

Too ambitious?

**Comment by chipbite** [comment]
Some other ideas for similar need: (1) Lookup (external to drawio, since it seems a bit involved to get a plugin going) of components, fill clipboard with drawio shape for the looked up component and paste it into editor. (2) Generate diagram skeleton, choosing components to add and then opening up in drawio to work on layout...

**Comment by davidjgraph** [comment]
I'd suggest to break these down into individual, distinct, well described feature requests, one per feature. Put them in issues, rather than discussion. I under the idea of discussing high level ideas here, but it needs specifics in issues in order to progress. Worse case is I say no :).

I know they might be inter-related, but in this form of discussion it's rather large and vague, which means a low change of implementation.

Specific use cases with pictures also helps.

**Comment by GanizaniSitara** [comment]
I've spent the last two weeks automating 3 level diagram (example in my repo) - the whole thing is naively written, but works very well. 

To synthesize what I think would be useful:

1) Create custom link action "send lookup" 
2) Allow it to define URL to send the lookup to and
2) Name of a defined property whose value it would pass to the lookup URL 

EDIT: the fact I'm now generating everything programmatically kind of does away with this feature, might still be useful for casual users if they have a service they could hook this up to. Then again, many casual users won't be building custom links. 

PS: drawio is absolutely fantastic :)

**Comment by GanizaniSitara** [comment]
So a simple feature would be.

I have a shape that defines a property, say "c4Name".

Make it referrable in the URL field say "http://myconfluence/%c4Name%" - which would get replace on click.

**Comment by chipbite** [comment]
@GanizaniSitara - sorry, I dont quite get your "simple feature" suggestion. 

> Make it referrable in the URL field say "http://myconfluence/%c4Name%" - which would get replace on click.

What do you mean with "the URL field" - would this action replace the standard link for a shape?

What is expected to be in the URL? A json file with exact same property/ies as the properties of the object?
What is getting replaced - one property or many?

What click would trigger a replace?

Also, how would you populate the shape with URLs? By hand? 


I am curious here, I have some ideas in this area.

**Reply by GanizaniSitara** [reply]
Hi, 
so, shape has a c4Name prorperty with value "MyApplication". The token replacement already works as a placeholder in label (described here https://www.diagrams.net/blog/placeholders)

I'd like to be able to use it in a standard hyperlink field with the same replacement value so that http://confluence/%c4Name% would become http://confluence/MyApplication on click.

One property would be sufficient to start with.

What I'd ultimately want to do is to have http://redirector/%c4Name% (or use any other property on other, non-C4 shapes). The redirector would lookup and translate historical and colloquial names of applications and components to the right form and land you on the appropriate link. It would also provide disambiguation on an external page like Wikipedia does (How many "Phoenix" projects does your enterprise have? There are at least 3 applications in ours, plus few projects, and I'm not even kidding).

**Reply by chipbite** [reply]
> I'd like to be able to use it in a standard hyperlink field with the same replacement value so that http://confluence/%c4Name% would become http://confluence/MyApplication on click.

So this UC is simply to allow placeholders in the link, same as in labels. I see.

Redirector: Yes, I see that UC as well. 

I am thinking about these things differently, we are using orbus iserver for enterprise architecture and for my department we are ofc exposing our work there, but are reluctant to having all solution architects work in iserver (they dont want to, and dont want to use visio). Also, it is rather closed, slow and cumbersome. Like most EA tools, it is closed and sucks.

So, I've been thinking about some simple EA-tool-like solutions, for a smaller context, , using a more open and devops slant, with drawio as the drawing component and format. But, nothing concrete yet...

**Reply by GanizaniSitara** [reply]
The EA tools angle is exactly the one I'm coming at it from. We're about to get rid of the one we have, which most EAs hate, and before we rush into paying for another we're looking at options that would be more open and user friendly.

**Reply by GanizaniSitara** [reply]
So I've done a little redirector that intelligently figures up where to send you, but still have to encode params into the URL by hand. I'm at a stage where reading the label would be very useful.
